### ISSUE TITLE
How to use a Personal Access Token?

### ISSUE THREAD

**Issue by ghost** [jenkinsci/stashnotifier-plugin#216]
See also: https://github.com/jenkinsci/stashnotifier-plugin/issues/24

It is now possible to authenticate against the [Bitbucket API with a Personal Access Token](https://confluence.atlassian.com/bitbucketserver059/personal-access-tokens-949254787.html?utm_campaign=in-app-help&utm_medium=in-app-help&utm_source=stash#Personalaccesstokens-usingpersonalaccesstokens).

How to do that using this plugin?

**Comment by scaytrase** [comment]
I think personal access tokens are not supported yet. PR is always welcome

**Comment by ghost** [comment]
@scaytrase I don't have experience creating/updating Jenkins plugins. 
And also I couldn't find a CONTRIBUTING.md or something similar in the repo?
So not sure where to start?

**Comment by scaytrase** [comment]
Actually we need to take care of the following code

Filling credentials select box:
https://github.com/jenkinsci/stashnotifier-plugin/blob/323b4778b49dff2a7efe9874884c561f37bb6aa0/src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java#L604-L630

Accessing credentials (basic auth is currently on the list)
https://github.com/jenkinsci/stashnotifier-plugin/blob/323b4778b49dff2a7efe9874884c561f37bb6aa0/src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java#L853-L892

There is bunch of code to utilize certificates as SSL auth, but it's declared non-working on #154 so I think it's not really relevant here

**Comment by ghost** [comment]
@scaytrase are you able to share any environment configuration setup guidelines? I'm on Windows. 

Doesn't seem too complicated to make the changes. Just trying to find out how to test the changes.

**Comment by scaytrase** [comment]
@rolandoldengarm 
I can't remebmer any specials in configuring local build env, I think all I use is IDEA, gradle and oracle jdk8

At last you can just open PR and HPI would be built for you automagickally (and even would be distributed with incrementals repo)

**Comment by artizirk** [comment]
Personal Access Tokens seem to work fine out of the box when testing with `curl`.

```sh
echo '{"state":"SUCCESSFUL", "description":"Test description", "key":"1234asdf", "url":"http://jenkins:8081/"}' | curl -X POST -H "Authorization: Bearer <token>" -H "Content-Type: application/json"  'https://<stash_server>/rest/build-status/1.0/commits/<commit_hash>' -d @- -v
```

**Comment by scaytrase** [comment]
Sure, it's just not supported for know here. So tell the labels

**Comment by darxriggs** [comment]
The [documentation](https://confluence.atlassian.com/bitbucketserver055/personal-access-tokens-940682155.html) states the following:
> For git operations, you can use your personal access tokens with your REST API.
>
> In addition to basic auth over REST API, you can use it as a bearer token, by setting the personal access token as a header value, instead of providing user name and password.

Therefore a personal access token can be used in the same way as a password in this plugin.
Create a username/password credential with the password field containing the personal access token.

@rolandoldengarm I suggest to try this and afterwards file a pull request with an updated _readme.md_ that states that and how to use username/password credentials for passwords and personal access tokens.

**Comment by noxo** [comment]
AFAIK "Personal Access Token" can't be used with current implementation which is using HTTP Basic Authentication, i.e it's populating username and password in "Authorization: Basic" header.

Implementation wound needs a bit tweaking, i.e start populating access token in "Authorization: Bearer" header.

**Comment by scaytrase** [comment]
Implemented in #276 released as 1.24
